internal/validation/ingress/ingress.go

```diff
@@ -0,0 +1,57 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ingress
+
+import (
+	"github.com/projectcontour/contour/internal/annotation"
+	networking_v1 "k8s.io/api/networking/v1"
+	"k8s.io/api/networking/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+)
+
+// DefaultClassName is the default IngressClass name that Contour will match
+// Ingress resources against if no specific IngressClass name is configured.
+const DefaultClassName = "contour"
+
+// MatchesIngressClassName returns true if the passed in Ingress annotations
+// or spec ingress class name matche the passed in ingress class name.
+// Annotations take precedence over spec field if both are set.
+func MatchesIngressClassName(obj metav1.Object, ingressClassName string) bool {
+	annotationClass := annotation.IngressClass(obj)
+	var specClass string
+	switch o := obj.(type) {
+	case *networking_v1.Ingress:
+		specClass = pointer.StringPtrDerefOr(o.Spec.IngressClassName, "")
+	case *v1beta1.Ingress:
+		specClass = pointer.StringPtrDerefOr(o.Spec.IngressClassName, "")
+	}
+
+	// If annotation is set, check if it matches.
+	if annotationClass != "" {
+		return annotation.MatchesIngressClass(obj, ingressClassName)
```

Comment:
WDYT about moving `annotation.MatchesIngressClass` into this package? I think it'd be nice to have all the ingress class matching logic in one place. I *think* it could even be unexported, and just let `MatchesIngressClassName` be the sole entrypoint for this package (but possible I'm overlooking something). 

I'm also very fine with doing this later, if we want to do it but it's gonna cause a bunch of churn right now while we're trying to get the release out.

Reply:
Yeah, I considered moving that but I think it is also used to check the annotation on HTTPProxy resources as well since we support setting a class in an annotation on those too.

I'll add an issue to clean this up b/c I think it could def be structured better

Reply:
Sounds good. Yeah it is used for `HTTPProxy` as well, but `MatchesIngressClassName` takes a `metav1.Object` which could be either an `Ingress` or an `HTTPProxy`, and we could type-switch as needed. Anyway we can discuss some more on the issue.

Reply:
https://github.com/projectcontour/contour/issues/3532

Reply:
Please comment and/or edit the issue if you think of some more ideas/context